Alternative paranoid mode

runtian-zhou opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Move Language.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

This is an alternative impl of #598. In this impl the type information are split out from the value.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

runtian-zhou commented on language/move-vm/types/src/loaded_data/annotated_types.rs:
```diff
@@ -0,0 +1,174 @@
+}
+
+#[derive(Debug, Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub enum ParanoidType {
```
This is the new separated type tag for values

runtian-zhou commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -732,19 +844,25 @@ const OPERAND_STACK_SIZE_LIMIT: usize = 1024;
-struct Stack(Vec<Value>);
+struct Stack {
+    value: Vec<Value>,
+    types: Vec<ParanoidType>,
```
Stacks now have a vector of types along side the value.

sblackshear replied:
Nice, this seems less disruptive than having a stack of (Value, Type) pairs

runtian-zhou commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -824,8 +982,10 @@ impl CallStack {
 struct Frame {
     pc: u16,
     locals: Locals,
+    local_tys: Vec<ParanoidType>,
```
Locals now carry a type as well.

sblackshear replied:
I think these can probably be read directly from `function.locals`

runtian-zhou commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -864,6 +1032,188 @@ impl Frame {
+            .ok_or_else(|| PartialVMError::new(StatusCode::VERIFIER_INVARIANT_VIOLATION))
+    }
+
+    fn type_stack_transition(
```
This is the main entrypoint for the paranoid checks.

tnowacki replied:
Yeah, this wasn't quite what I was expecting, though this may be easier? I'm not sure. 

I was imagining replacing all interactions with the stacks with macros 

`pop` and `push` and `popn`/`pushn`/`lastn`. 

This way you could do something like `let (ty, value) = pop!(interpreter)` And even swap out the type for `ty` with a `cfg` feature flag if we want to disable the paranoid mode

tnowacki commented on language/move-vm/runtime/src/loader.rs:
```diff
@@ -2032,6 +2188,7 @@ impl Function {
             type_parameters,
             native,
             def_is_native,
+            def_is_friend_or_private,
```
I haven't looked where this is used, but wouldn't you want to know the exact visibility?

tnowacki commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -246,6 +273,55 @@ impl Interpreter {
+        if callee.is_friend_or_private() {
+            match (caller.module_id(), callee.module_id()) {
+                (Some(caller_id), Some(callee_id)) => {
+                    if caller_id.address() == callee_id.address() {
```
This isn't inherent to friends. Like sure we have this check in friends right now, but it might not always be the case. I would probably check if the module is in the friend list?

It also doesn't catch invalid private access from other addresses

runtian-zhou commented:
As mentioned, we would prefer this implementation over the one in #598. Thus I'll solicit the review here. cc @sblackshear @tnowacki @vgao1996

vgao1996 commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -251,18 +271,28 @@ impl Interpreter {
+        &mut self,
+        func: Arc<Function>,
+        ty_args: Vec<Type>,
+        paranoid: bool,
```
I think you can directly access `self.runtime_config.paranoid_type_checks` without passing the flag in, no?

vgao1996 commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -311,6 +341,20 @@ impl Interpreter {
         }
+
+        if self.runtime_config.paranoid_type_checks {
+            let is_generic = !ty_args.is_empty();
```
Nit: you don't have to branch based on whether the function is generic or not. Just call `function.parameter_types()[expected_args - i - 1].subst(&ty_args)?` and it'll work for both generic and non-generic functions -- if `ty_args` is empty, then the substitution is effectively equivalent to a clone.

vgao1996 replied:
> efficiency of subst

No need to worry about this. `subst`'s implementation is near identical to clone.

vgao1996 commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -356,9 +400,52 @@ impl Interpreter {
         }
+
+        if self.runtime_config.paranoid_type_checks {
+            let is_generic = !ty_args.is_empty();
```
Similar here. No need to differentiate the two cases.

vgao1996 reviewed (approved):
Overall looks good to me. I like this implementation! Glad it turns out much simpler than the original approach.

Let's still leave @sblackshear @tnowacki and maybe @jolestar some time to take a look

jolestar reviewed (commented):
I like this solution. 

I have an idea, maybe we can check the type in `pop_as<T>`, and make sure the type in the stack matches the T.

Then we can eliminate `simple_instruction_type_stack_transition` and handle the instruction type and value together.

sblackshear commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -869,6 +1033,504 @@ impl Frame {
+                        .remove(Ability::Key)
+                        .union(AbilitySet::singleton(Ability::Store))
+                } else {
+                    ability
```
It took me a minute to convince myself that this is correct in the case that the struct being packed has `store`, but not `key`? e.g.,
```
struct A {}
struct B has store {
  a: A
}

let x = B { a: A{} } // should get flagged by runtime checks
```

A comment on why this is safe might be helpful.

sblackshear commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -869,6 +1033,504 @@ impl Frame {
+                let ty = resolver.get_struct_type(*idx);
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::MutableReference(Box::new(ty)))?;
```
I think we want to check for the `key` ability in all of the global storage operators. The bytecode verifier does this, anway.

sblackshear commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -869,6 +1033,504 @@ impl Frame {
+                        interpreter.operand_stack.push_ty(Type::Reference(ty))?
+                    }
+                    Type::Reference(ty) => {
+                        interpreter.operand_stack.push_ty(Type::Reference(ty))?
```
I think we want this to be an error. The verifier will not allow you to apply `FreezeRef` to an immutable reference.

runtian-zhou replied:
There's a very interesting observation here: it seems that for vector instructions we allow implicit reference casting (mut reference can be used as immut reference). e.g: VecLen

sblackshear reviewed (commented):
Like what you've done with this approach Runtian! Flagged a few bugs (I think, `StoreLoc` being the most high-consequence one) and a few cosmetic items.

runtian-zhou commented:
> I have an idea, maybe we can check the type in pop_as<T>, and make sure the type in the stack matches the T.

I tried that previously. The api would look more like

`fn pop_with_checks<F>(&self, check_fn: F) where F: FnOnce(Type) -> PartialVMResult<()>`

It worked but it was just quite ugly as the type transition was tangled with value pops and the code didn't seem as clean as this version.

runtian-zhou commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -869,6 +1033,504 @@ impl Frame {
+            | Bytecode::And => {
+                let lhs = interpreter.operand_stack.pop_ty()?;
+                let rhs = interpreter.operand_stack.pop_ty()?;
+                lhs.check_eq(&rhs)?;
```
So we don't check the types for primitive value operations, including integer operations, boolean operations and comparision operations. The reason why they were safe is because our value system have separate representation between primitive values and struct values, so had there been a type confusion they would be caught at the runtime. I can either add a test to make sure it doesn't regress or I can add the checks here. What would be your preference?

sblackshear replied:
Either sounds great to me--as long as something will break if we switch to a different value representation in the future. Thanks!

runtian-zhou commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -150,12 +160,16 @@ impl Interpreter {
                 ExitCode::Call(fh_idx) => {
                     let func = resolver.function_from_handle(fh_idx);
 
+                    if self.runtime_config.paranoid_type_checks {
```
Same. We don't have `config` for `Frame`.. Sry it was a bit confusing.

sblackshear commented on language/move-vm/runtime/src/interpreter.rs:
```diff
@@ -869,6 +1033,504 @@ impl Frame {
+                interpreter.operand_stack.pop_ty()?;
+            }
+            // StLoc needs to check before execution as we need to check the drop ability of values.
+            Bytecode::StLoc(idx) => {
```
I see, I did not realize there were two sets of checks. What would go wrong if we did all the checks either before execution or after (minus the gas issue mentioned on "before", which seems fairly minor)

runtian-zhou replied:
For some of the operations, the cost of resolving the types seems non trivial to me. That's the main reason that makes me feel that we should only do the checks after we've charged proper gas here. Would be great if @vgao1996 can weigh in your gas experience here. To answer your question:

- Calling everything before will work, but just that we might run the checks before we charge gas, and some of those checks could be more expensive. 
- Calling everything after won't work for all control flow operators and `StLoc`. The reason is that for control flow oeprators the after path won't be invoked at all. The `StLoc` is a very interesting case here. The reason why it won't work is because we need to check if the local we can potentially overwrite has the drop ability, and the only way to check that is before we execute the instruction.

sblackshear replied:
Thanks for the explanation. My bias would be for clarity/readability--these checks should only be run without charging gas in the (hopefully rare/impossible) event that that there's a bug in the bytecode verifier, and the cost of any individual instruction is (hopefully) very small. So I think the scenario that splitting the checks is protecting against is both unlikely and low consequence. Whereas a bug in the checks that happens due to the before/after split would be unlikely and high consequence.

This is subjective and I'm ok with whatever you decide on this.

sblackshear commented:
> There's a very interesting observation here: it seems that for vector instructions we allow implicit reference casting (mut reference can be used as immut reference). e.g: VecLen

Yep, and same for (e.g.) BorrowFieldImmut. Only function calls require strict reference type equality.

runtian-zhou commented:
Had a chat with @vgao1996 and we would land the PR in the current form, i.e: run the checks at two locations: pre instruction execution and post instruction execution. The reason is that we want to run as less code as possible before we charge proper gas. I updated the comment to address the gas concern here. To address @sblackshear's concern on making sure the two functions constitutes the full type safety checks I updated the code to use exhaustive match to make sure we cannot miss checks in the future if we decide to add new instructions.

Thanks for everyone for looking into this! Added a few comments and I'll merge this PR shortly.
